File: Day01/part2.py
file_path = "input.txt"

dial = 50
password = 0

try:
	with open(file_path, 'r') as file:
		for line in file:
			line = line.strip()
			if len(line) < 2:
				continue
			clicks = int(line[1:])
			if line[0] == 'L':
				if dial == 0:
					password += clicks // 100
				elif clicks < dial:
					pass
				else:
					clicks -= dial
					password += 1
					password += clicks // 100
				dial = (dial - int(line[1:])) % 100
			else:
				if dial == 0:
					password += clicks // 100
				elif clicks < 100 - dial:
					pass
				else:
					clicks -= 100 - dial
					password += 1
					password += clicks // 100
				dial = (dial + int(line[1:])) % 100
	
	print(f"The password is: {password}")
except Exception as e:
	print(f"An error occurred: {e}")

File: Day05/part2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#define ll long long int



int main(){
	std::ifstream fin("input.txt");
	std::string line;

	if (!fin.is_open()) {
		std::cout << "Error: Cannot open file!" << std::endl;
		return 1;
	}

	// pair of id and whether it is the left side of range
	std::vector<std::pair<ll, bool>> points;
	ll count = 0;

	while (std::getline(fin, line)) {
		if (line.length() == 0) {
			break;
		}

		size_t hyphenPos = line.find('-');
		ll leftRange = std::stoll(line.substr(0, hyphenPos));
		ll rightRange = std::stoll(line.substr(hyphenPos+1));
		
		points.push_back(std::make_pair(leftRange, true));
		points.push_back(std::make_pair(rightRange, false));
	}

	std::sort(points.begin(), points.end(), [](const std::pair<ll, bool>& a, const std::pair<ll, bool>& b) {
		if (a.first != b.first) return a.first < b.first;
		else return a.second > b.second;
	});
	points.insert(points.begin(), points[0]);


	int inRange = 0;
	for (std::vector<std::pair<ll, bool>>::iterator i = points.begin() + 1; i != points.end(); ++i) {
		if (inRange == 0) {
			if (i->second) {
				inRange++;
			} else {
				inRange--;
			}
			count++;
			continue;
		}
		if (i->second) {
			inRange++;
		} else {
			inRange--;
		}

		count += i->first - (i-1)->first;
	}



	fin.close();

	std::cout << "The answer is: " << count << std::endl;

	return 0;
}
